cpp/dolfinx/io/VTKHDF.h

```diff
@@ -215,36 +249,31 @@ mesh::Mesh<U> io::VTKHDF::read_mesh(MPI_Comm comm, std::string filename)
                 comm);
   std::partial_sum(recv_count.begin(), recv_count.end(),
                    recv_offsets.begin() + 1);
-  std::vector<std::uint8_t> recv_types(recv_offsets.back());
-  MPI_Allgatherv(types_unique.data(), count, MPI_UINT8_T, recv_types.data(),
-                 recv_count.data(), recv_offsets.data(), MPI_UINT8_T, comm);
+
+  MPI_Datatype compound_type;
+  MPI_Type_contiguous(2, MPI_UINT8_T, &compound_type);
+  MPI_Type_commit(&compound_type);
+
+  std::vector<std::array<std::uint8_t, 2>> recv_types(recv_offsets.back());
```

Comment:
This look very wrong. You need to use a flat `std::vector` with length `2 * recv_offsets.back()`.

Passing `recv_types.data()` is a pointer to the first `std::array`, rather than a pointer to the first `std::uint8_t`